### PR TITLE
Camera model support

### PR DESCRIPTION
**Pull request by DSaurus**
Scene could add multiple cameras to render.
But examples should be changed because of the new camera-based render method.
if you think this method is reasonable, I will modify other examples to make them work.

### CODE REVIEW

File: taichi_three/geometry.py [hunk 1]
```diff
@@ -11,48 +11,48 @@ def render_triangle(model, face):
...
-    B_A *= ilB_A
-    C_B *= ilC_B
-    A_C *= ilA_C
-    BxA = ts.cross(B, A) * ilB_A
-    CxB = ts.cross(C, B) * ilC_B
-    AxC = ts.cross(A, C) * ilA_C
-    normal = ts.normalize(ts.cross(a - c, a - b))
-    light_dir = scene.camera.untrans_dir(scene.light_dir[None])
-    center_pos = (a + b + c) / 3
+    for camera in ti.static(scene.cameras):
```
Comment (archibate) [inline taichi_three/geometry.py:14]: Can we use
`def render_triangle(camera, model, face):`
and move this for-loop to the caller?

Comment (DSaurus) [reply taichi_three/geometry.py:14]: Yes, I will move it.

File: taichi_three/geometry.py [hunk 2]
```diff
@@ -47,12 +47,17 @@ def render_triangle(model, face):
...
-
-            if zindex >= ti.atomic_max(scene.zbuf[X], zindex):
+        if X.x < 0 or X.x >= camera.res[0] or X.y < 0 or X.y >= camera.res[1]:
+            continue
+        if AB >= 0 and BC >= 0 and CA >= 0:
+            w_A = max(Ak * BC, 1e-6)
+            w_B = max(Bk * CA, 1e-6)
+            w_C = max(Ck * AB, 1e-6)
+            w_sum = w_A + w_B + w_C
+            zindex = 1.0 /  ( (a.z * w_A + b.z * w_B + c.z * w_C) / w_sum)
```
Comment (archibate) [inline taichi_three/geometry.py:57]:
```suggestion
            zindex = w_sum / (a.z * w_A + b.z * w_B + c.z * w_C)
```

File: taichi_three/loader.py [hunk 3]
```diff
@@ -62,6 +62,15 @@ def readobj(path, scale=None):
...
 
+    if vi.__len__() == 0:
+        vi.append([0, 0, 0])
+    if vt.__len__() == 0:
+        vt.append([0, 0])
+    if vn.__len__() == 0:
+        vn.append([0, 0, 0])
+    if faces.__len__() == 0:
+        faces.append([0, 0, 0])
+
```
Comment (archibate) [inline taichi_three/loader.py:73]: Why these? `len(faces)` should also work.

Comment (DSaurus) [reply taichi_three/loader.py:73]: In model.py, but vx_n may never be None if use objloader.
```python
        if vi_n is None:
            vi_n = 1
        if vt_n is None:
            vt_n = 1
        if vn_n is None:
            vn_n = 1
```
When vn_n is zero, this part will throw an error:
```python
        if f_n is not None:
            self.faces = ti.Matrix.var(3, f_m, ti.i32, f_n)
        if vi_n is not None:
            self.vi = ti.Vector.var(3, ti.f32, vi_n)
        if vt_n is not None:
            self.vt = ti.Vector.var(2, ti.f32, vt_n)
        if vn_n is not None:
            self.vn = ti.Vector.var(3, ti.f32, vn_n)
        if tex_n is not None:
            self.tex = ti.Vector.var(3, ti.f32, tex_n)
```
I will write something better to fix this bug.

**Comment by archibate** [comment]
I found that the x-axis of `L2W.from_mouse(gui)` seems reversed?

**Comment by DSaurus** [comment]
> I found that the x-axis of `L2W.from_mouse(gui)` seems reversed?

Maybe the reason is that Z-buffer is "max 1/z" Z-buffer now, it is reversed compared to before. ("max z" Z-buffer)
And because of that, I reverse the camera position z and model position z setting in "from mouse", but I don't modify position x so the x-axis of `L2W.from_mouse(gui)` is reversed.

**Pull request merged**